### PR TITLE
Update network section in 2.5 porting guide

### PR DESCRIPTION
**Pull request by ganeshrn**
##### SUMMARY

*  Mention of refactoring Network shared code found
   in module_utils into a new package.

##### ISSUE TYPE

 - Docs Pull Request

##### COMPONENT NAME

rst/porting_guide_2.5.rst

##### ANSIBLE VERSION

```
2.5
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: docs/docsite/rst/porting_guide_2.5.rst [hunk 1]
```diff
@@ -209,3 +209,11 @@ Using a provider dictionary with one of the new persistent connection types for
 (network_cli, netconf, etc.) will result in a warning. When using these connections
 the standard Ansible infrastructure for controlling connections should be used.
 (Link to basic inventory documentation?)
+
+Developers:
```
Comment (gundalow) [inline docs/docsite/rst/porting_guide_2.5.rst]:
```
Developers 
----------
```

File: docs/docsite/rst/porting_guide_2.5.rst [hunk 2]
```diff
@@ -209,3 +209,11 @@ Using a provider dictionary with one of the new persistent connection types for
...
+
+* The network specific shared code found in 'ansible.module_utils' is moved to 'ansible.module_utils.network'.
```
Comment (gundalow) [inline docs/docsite/rst/porting_guide_2.5.rst]: In rst it's
```
``module_utils.network``
````

File: docs/docsite/rst/porting_guide_2.5.rst [hunk 3]
```diff
@@ -209,3 +209,11 @@ Using a provider dictionary with one of the new persistent connection types for
...
+* Going ahead all the Network platform specific code will reside in 'ansible.module_uitls.network.{{ platform }}'
+  where {{ platform }} is platform specific shared functions. The network common shared functions which are
+  platform independent will reside in 'ansible.module_utils.network.common'. This is noted in the module utilities
+  developer guide.
```
Comment (gundalow) [inline docs/docsite/rst/porting_guide_2.5.rst]: Which guide, we should link to it

File: docs/docsite/rst/porting_guide_2.5.rst [hunk 4]
```diff
@@ -209,3 +209,11 @@ Using a provider dictionary with one of the new persistent connection types for
 (network_cli, netconf, etc.) will result in a warning. When using these connections
 the standard Ansible infrastructure for controlling connections should be used.
 (Link to basic inventory documentation?)
+
+Developers:
+
+* The network specific shared code found in 'ansible.module_utils' is moved to 'ansible.module_utils.network'.
+* Going ahead all the Network platform specific code will reside in 'ansible.module_uitls.network.{{ platform }}'
```
Comment (gundalow) [inline docs/docsite/rst/porting_guide_2.5.rst]: Suggestion: From Ansible 2.5, onwards, the Network platform specific code..

**Comment by gundalow** [comment]
Assigning to @acozine for final review and merge.

File: docs/docsite/rst/porting_guide_2.5.rst (hunk 3 above)
Comment (ganeshrn) [reply docs/docsite/rst/porting_guide_2.5.rst]: The details of file are mentioned here https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/dev_guide/developing_module_utilities.rst

**Comment by ganeshrn** [comment]
ready_for_review

**Review by gundalow — APPROVED** [review]
I'm happy with technical content, will let @acozine wordsmith.

File: docs/docsite/rst/porting_guide_2.5.rst [hunk 5]
```diff
@@ -209,3 +209,12 @@ Using a provider dictionary with one of the new persistent connection types for
 (network_cli, netconf, etc.) will result in a warning. When using these connections
 the standard Ansible infrastructure for controlling connections should be used.
 (Link to basic inventory documentation?)
+
+Developers
+----------
+
+* The network specific shared code found in ``ansible.module_utils`` is moved to ``ansible.module_utils.network``.
+* From Ansible 2.5, onwards, the Network platform specific code will reside in ``ansible.module_uitls.network.{{ platform }}``
```
Comment (acozine) [inline docs/docsite/rst/porting_guide_2.5.rst]: typo here - `uitls` instead of `utils`

**Comment by acozine** [comment]
Thanks for contributing @ganeshrn - I'm going to use this PR as an opportunity to pick your and @gundalow's brains a bit.

I like the idea of Porting Guides and would like to know more. Are they based on release notes? What's the audience for the Porting Guides? What are the goals? 

For this particular change, an active description with examples would make it stronger. And here's where I need your help, since I don't actually know what a developer's use of one of these tools looks like.

Here's a guesswork/nonsense example: 
"Shared code for network modules has moved to `ansible.module_utils.network` beginning in Ansible 2.5. Platform-independent functions are in `ansible.module_utils.network.common`; platform-specific functions are in `ansible.module_utils.network.{{ platform }}`. If your module uses the shared code from `ansible.module_utils`, you must update all references. For example, change 
`call ansible.module_utils.router_backscratcher` to `call ansible.module_utils.network.router_backscratcher` 
and change `call ansible.module_utils.vyos_router_magic` to `call ansible.module_utils.network.vyos.vyos_router_magic`. For more information ,see the Developer Guide."

So . . . what would the real examples look like? How does this change affect developers?

**Comment by acozine** [comment]
Also, does this change affect all developers, or only some? In other words, does everyone use the shared functions?

**Comment by gundalow** [comment]
I can talk through the Roadmaps, changelog and Porting guide outside of this PR.

This particular entry is for developers.

**Comment by gundalow** [comment]
Example changes needed
In a module (network/eos/eos_banner.py)
https://github.com/ganeshrn/ansible/commit/f42ff8ae869bd03b2d2746291f113efc9932ff61#diff-1fd7bf472949eba0f39bdbcb527eeefc


Module_utils that have moved: https://github.com/ganeshrn/ansible/commit/f42ff8ae869bd03b2d2746291f113efc9932ff61#diff-1ef5433713e35f651da836a9db4545b5L10

**Comment by gundalow** [comment]
Affects all people developing modules. Any module that ships with ansible has been updated. This is more of a heads up for people working on modules that haven't been upstreamed (PRs in flight) OR people that maintain their own modules in their company repos.

**Comment by gundalow** [comment]
Not sure how far we want to go with this. We *could* document how to write modules that work with 2.4 and 2.5, along the lines of https://github.com/F5Networks/f5-ansible/blob/devel/library/bigip_config.py#L121-L147

File: docs/docsite/rst/porting_guide_2.5.rst (hunk 5 above)
Comment (ganeshrn) [reply docs/docsite/rst/porting_guide_2.5.rst]: Fixed

**Comment by gundalow** [comment]
I'm happy with the content, with @acozine for final review and merge

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
In the module. such as `modules/network/vyos_banner.py`
**Before (2.4)**
```
from ansible.module_utils.vyos import get_config, load_config
from ansible.module_utils.vyos import vyos_argument_spec
```

**After (2.5)**
```
from ansible.module_utils.network.vyos.vyos import get_config, load_config
from ansible.module_utils.network.vyos.vyos import vyos_argument_spec
```

**Comment by acozine** [comment]
Thanks @ganeshrn for documenting this, and thanks @gundalow for the examples.

**Comment by gundalow** [comment]
Looks great, thanks!

**Pull request merged**